src/Surfnet/StepupSelfService/SelfServiceBundle/Controller/Registration/GssfController.php

```diff
@@ -0,0 +1,181 @@
+    }
+
+    /**
+     * @param Request $httpRequest
+     * @param string  $provider
+     * @return array|Response
+     */
+    public function consumeAssertionAction(Request $httpRequest, $provider)
+    {
+        $provider = $this->getProvider($provider);
+
+        $this->get('logger')->notice(
+            sprintf('Received SAMLResponse from GSSP "%s", attempting to process', $provider->getName())
+        );
+
+        try {
+            /** @var \Surfnet\SamlBundle\Http\PostBinding $postBinding */
+            $postBinding = $this->get('surfnet_saml.http.post_binding');
+            $assertion = $postBinding->processResponse(
+                $httpRequest,
+                $provider->getRemoteIdentityProvider(),
+                $provider->getServiceProvider()
+            );
+        } catch (Exception $exception) {
+            $this->getLogger()->error(
+                sprintf('Could not process received Response, error: "%s"', $exception->getMessage())
+            );
+
+            return $this->render(
+                'SurfnetStepupSelfServiceSelfServiceBundle:Registration/Gssf:initiate.html.twig',
+                ['provider' => $provider->getName(), 'authenticationFailed' => true]
+            );
+        }
+
+        $adaptedAssertion = new AssertionAdapter($assertion);
+        $expectedResponseTo = $provider->getStateHandler()->getRequestId();
+
+        if (!$adaptedAssertion->inResponseToMatches($expectedResponseTo)) {
+            $this->getLogger()->critical(sprintf(
+                'Received Response with unexpected InResponseTo: "%s", %s',
+                $adaptedAssertion->getInResponseTo(),
+                ($expectedResponseTo ? 'expected "' . $expectedResponseTo . '"' : ' no response expected')
+            ));
+
+            return $this->render(
+                'SurfnetStepupSelfServiceSelfServiceBundle:Registration/Gssf:initiate.html.twig',
+                ['provider' => $provider->getName(), 'authenticationFailed' => true]
+            );
+        }
+
+        $this->get('logger')->notice(
+            sprintf('Processed SAMLResponse from GSSP "%s" successfully', $provider->getName())
+        );
+
+        /** @var \Surfnet\StepupSelfService\SelfServiceBundle\Service\GssfService $service */
+        $service = $this->get('surfnet_stepup_self_service_self_service.service.gssf');
+        /** @var \Surfnet\SamlBundle\SAML2\Attribute\AttributeDictionary $attributeDictionary */
+        $attributeDictionary = $this->get('surfnet_saml.saml.attribute_dictionary');
+        $gssfId = $attributeDictionary->translate($assertion)->getNameID();
+
+        if ($secondFactorId = $service->provePossession($this->getIdentity()->id, $provider->getName(), $gssfId)) {
+            return $this->redirectToRoute(
+                'ss_registration_email_verification_email_sent',
+                ['secondFactorId' => $secondFactorId]
+            );
+        }
+
+        return $this->render(
```

Comment:
For all exitpoints: please don't forget to clear the state of the current authnrequest to the current GSSP (from the top of my head: `$provider->getStateHandler()->clear()`)